src/lib/comms/sol-bluetooth.c

```diff
@@ -0,0 +1,123 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+
+#include <sol-bluetooth.h>
+#include <sol-buffer.h>
+#include <sol-log.h>
+#include <sol-util.h>
+
+static int
+string_to_uuid128(struct sol_bt_uuid *uuid, const char *str)
+{
+    uint32_t data0, data4;
+    uint16_t data1, data2, data3, data5;
+    uint8_t *val = uuid->val128;
+
+    if (sscanf(str, "%08x-%04hx-%04hx-%04hx-%08x%04hx",
+            &data0, &data1, &data2,
+            &data3, &data4, &data5) != 6)
+        return -EINVAL;
+
+    data0 = sol_util_be32_to_cpu(data0);
+    data1 = sol_util_be16_to_cpu(data1);
+    data2 = sol_util_be16_to_cpu(data2);
+    data3 = sol_util_be16_to_cpu(data3);
+    data4 = sol_util_be32_to_cpu(data4);
+    data5 = sol_util_be16_to_cpu(data5);
+
+    memcpy(&val[0], &data0, 4);
+    memcpy(&val[4], &data1, 2);
+    memcpy(&val[6], &data2, 2);
+    memcpy(&val[8], &data3, 2);
+    memcpy(&val[10], &data4, 4);
+    memcpy(&val[14], &data5, 2);
+
+    return 0;
+}
+
+SOL_API int
+sol_bt_uuid_from_str(struct sol_bt_uuid *uuid, const char *str)
+{
+    size_t len;
+    int r;
+
+    SOL_NULL_CHECK(str, -EINVAL);
+    SOL_NULL_CHECK(uuid, -EINVAL);
+
+    len = strlen(str);
+
+    switch (len) {
+    case 4:
+        uuid->type = SOL_BT_UUID_TYPE_16;
+        uuid->val16 = strtol(str, NULL, 16);
+        break;
+    case 8:
+        uuid->type = SOL_BT_UUID_TYPE_32;
+        uuid->val32 = strtoll(str, NULL, 16);
+        break;
+    case 36:
+        uuid->type = SOL_BT_UUID_TYPE_128;
+        r = string_to_uuid128(uuid, str);
+        SOL_INT_CHECK(r, < 0, -EINVAL);
+        break;
+    default:
+        return -EINVAL;
+    }
+
+    return 0;
+}
+
+SOL_API int
+sol_bt_uuid_to_str(const struct sol_bt_uuid *uuid, struct sol_buffer *buffer)
+{
+    uint32_t tmp1, tmp5;
+    uint16_t tmp0, tmp2, tmp3, tmp4;
+    int r;
+
+    SOL_NULL_CHECK(uuid, -EINVAL);
+    SOL_NULL_CHECK(buffer, -EINVAL);
+
+    switch (uuid->type) {
+    case SOL_BT_UUID_TYPE_16:
+        r = sol_buffer_append_printf(buffer, "%.4x", uuid->val16);
+        break;
+    case SOL_BT_UUID_TYPE_128:
+        memcpy(&tmp0, &uuid->val128[0], sizeof(tmp0));
+        memcpy(&tmp1, &uuid->val128[2], sizeof(tmp1));
+        memcpy(&tmp2, &uuid->val128[6], sizeof(tmp2));
+        memcpy(&tmp3, &uuid->val128[8], sizeof(tmp3));
+        memcpy(&tmp4, &uuid->val128[10], sizeof(tmp4));
+        memcpy(&tmp5, &uuid->val128[12], sizeof(tmp5));
+
+        r = sol_buffer_append_printf(buffer , "%.8x-%.4x-%.4x-%.4x-%.8x%.4x",
```

Comment:
ah, I've missed that. I guessed ".8" would limit to 8 digits, but would pad with spaces. Just tried and confirmed it does 0-padding.